### ISSUE TITLE
Performance with --llvm lags for prk-stencil with --no-local

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#8060]
For prk-stencil, llvm performs on par with the gnu C backend with `--local`, but performance falls off for `--no-local`, which is what causes 16-node-xc prk-stencil to lag behind with `--llvm`

Here's a small reproducer that is just a stripped down version of prk-stencil (it can probably be stripped down further, but this is a good start):

```chpl
use Time, StencilDist;

config const order: int = 20000;
config param R = 2;

param Wsize = 2*R + 1,
              R1 = R + 1;

proc main() {
  const localDom = {0.. # order, 0.. # order},
   innerLocalDom = localDom.expand(-R);

  const outDist   = new dmap(new Stencil(innerLocalDom));
  const fluffDist = new dmap(new Stencil(innerLocalDom, fluff=(R,R)));

  const Dom = localDom dmapped fluffDist,
   innerDom = innerLocalDom dmapped fluffDist;

  const outputDom = localDom dmapped outDist;

  var input: [Dom] real = 0.0,
      output: [outputDom] real = 0.0;

  var weight: Wsize*(Wsize*(real));

  var timer: Timer; timer.start();
  //forall (i,j) in innerDom with (const in weight) {
  for (i,j) in innerDom {
    var tmpout: real = 0.0;
    for param jj in -R..-1 do tmpout += weight[R1][R1+jj] * input.localAccess[i, j+jj];
    for param jj in 1..R   do tmpout += weight[R1][R1+jj] * input.localAccess[i, j+jj];
    for param ii in -R..-1 do tmpout += weight[R1+ii][R1] * input.localAccess[i+ii, j];
    for param ii in 1..R   do tmpout += weight[R1+ii][R1] * input.localAccess[i+ii, j];
    output.localAccess[i, j] += tmpout;
  }
  writeln(timer.elapsed());
}
```

The generated code is actually pretty similar for `--local` and `--no-local`. The main difference is that the `--no-local` code has an extra dereference to get the `addr` out of the widened input array. 

Here's some performance numbers:

| config                                                         | time  |
| -------------------------------------------------------------- | ----- |
| chpl f.chpl --fast --no-local                                  | 2.05s |
| chpl f.chpl --fast --no-local --target-compiler=clang-included | 5.15s |
| chpl f.chpl --fast --no-local --llvm                           | 5.15s |
| chpl f.chpl --fast --no-local  --ccflags -fno-strict-aliasing  | 6.20s |

The interesting things to note are that clang-included and clang perform as poorly as `--llvm` and that gcc performs poorly with `-fno-strict-aliasing`. I'm hoping that @dmk42's stuct alias work will help here, but we'll have to see when that comes online.

Looking at the generated code you can see that our LICM is ineffective. Reverting 538386e6f7 allows all the meta-data to be hoisted, and clang-included/llvm outperform the C backend. That doesn't help the full stencil-prk, but I would guess there end up being atomics from the RadCache the prevent hoisting or because of the getPrivatizedCopy calls, which we could almost certainly workaround.

**Comment by ronawho** [comment]
I was curious to see if things improved at all with the [llvm 6 upgrade](https://github.com/chapel-lang/chapel/pull/8869)

Unfortunately (but not surprisingly) that didn't seem to help. For 16-node optimized prk-stencil:

| config                                                | GFlops/s |
| ----------------------------------------------------- | -------- |
| chpl f.chpl --fast --no-local                         |      705 |
| chpl f.chpl --fast --no-local --llvm                  |      400 |
| chpl f.chpl --fast --no-local --llvm --llvm-wide-opt  |      330 |

**Comment by ronawho** [comment]
After some more digging as part of https://github.com/chapel-lang/chapel/issues/9224, I believe improving LICM (to hoist getPrivatizedCopy calls -- https://github.com/chapel-lang/chapel/issues/6184) is currently the only practical way to address this.

I rebased on old experimental (and extremely incorrect/wrong) branch that allows hoisting of the `chpl_getPrivatizedClass` from parallel constructs: https://github.com/chapel-lang/chapel/compare/master...ronawho:licm-improve

 The basic changes are: 
 - inline `chpl_getPrivatizedCopy` (so LICM can see the `chpl_getPrivatizedClass` primitive)
 - update LICM to consider the `chpl_getPrivatizedClass` primitive as pure
 - disable aliasing checks that prevent hoisting of references and argbundles that get passed by ref to functions (this is the extremely incorrect part)

This allows array meta-data including  calls to `chpl_getPrivatizedClass()` to be hoisted, which allows the llvm backend to get good performance. For 16 node prk-stencil-opt:

| branch       | C            | llvm             |
| ------------ | ------------ | ---------------- |
| master       | 720 GFlops/s |   540 GFlops/s   |
| licm-improve | 720 GFlops/s | **720 GFlops/s** |


Unfortunately I think it will be a lot of work to improve LICM analysis to properly detect that these things are hoist-able. I'm guessing I'd need at least a month to do it. I'll need to:
 - re-familiarize myself with the compiler and LICM algorithm/implementation 
 - cleanup/improve the LICM code -- https://github.com/chapel-lang/chapel/issues/6411
 - actually implement the new analysis (this is the real effort, and would involve a lot of updates to our def/use and alias analysis)

**Comment by mppf** [comment]
It looks to me like indeed the global variable chpl_privateObjects is not being hoisted from the loop.

I wonder if LLVM has an optimization we can enable that detects which functions modify which globals...

**Comment by mppf** [comment]
Elliot pointed me also to #6411 and #6388. I'm thinking about making an LLVM pass that hoists chpl_getPrivatizedClass always and loop headers as long as the body of the loop doesn't call functions that modify loop headers.

**Comment by mppf** [comment]
I've been looking at using llvm.invariant.start to tell the LLVM LICM that the load of chpl_privateObjects can be hoisted. However that doesn't seem to be working for some reason I can't understand. A simple example with a similar pattern is working... but in the reproducer it's not.

**Comment by mppf** [comment]
I have something like this
```
%22 = load i8**, i8*** @chpl_privateObjects, align 8
```

in a loop where @chpl_privateObjects is marked invariant. But the LLVM LICM code really wants to find a bitcast from the operand, and that's not present in our code:

```
  while (Addr->getType() != PtrInt8Ty) {
    auto *BC = dyn_cast<BitCastInst>(Addr);
    // Avoid traversing high number of bitcast uses.
    if (++BitcastsVisited > MaxNumUsesTraversed || !BC)
      return false;
    Addr = BC->getOperand(0);
  }
```

We're running into that 'return false' because Addr is a global rather than a BitCastInst.

Perhaps copying the global into a local variable/bitcasting it will work. Or improving LICM.

**Comment by mppf** [comment]
Here is my reproducer that doesn't reproduce it yet:
```
; RUN: opt < %s -licm -S | FileCheck %s
; RUN: opt < %s -aa-pipeline=basic-aa -passes='require<opt-remark-emit>,loop(licm)' -S | FileCheck %s

@chpl_privateObjects = external dso_local global i8**, align 8

declare {}* @llvm.invariant.start.p0i8(i64, i8* nocapture) nounwind readonly
declare void @llvm.invariant.end.p0i8({}*, i64, i8* nocapture) nounwind

define i8 @test_licm_global(i32 %n) {
; CHECK-LABEL: @test_fence
; CHECK-LABEL: entry
; CHECK: invariant.start
; CHECK: %addrld = load atomic i32, i32* %addr.i unordered, align 8
; CHECK: br label %loop
entry:
  %invst = tail call {}* @llvm.invariant.start.p0i8(i64 8, i8* bitcast (i8*** @chpl_privateObjects to i8*)) #4
  br label %loop

loop:
  %indvar = phi i32 [ %indvar.next, %loop ], [ 0, %entry ]
  %sum = phi i8 [ %sum.next, %loop ], [ 0, %entry ]

  %g = load i8**, i8*** @chpl_privateObjects, align 8
  %p = load i8*, i8** %g, align 8
  %v = load i8, i8* %p, align 8

  %sum.next = add i8 %v, %sum
  %indvar.next = add i32 %indvar, 1
  %cond = icmp slt i32 %indvar.next, %n
  br i1 %cond, label %loop, label %loopexit

loopexit:
  ret i8 %sum
}

attributes #4 = { nounwind }
```

**Comment by ronawho** [comment]
> I'm thinking about making an LLVM pass that hoists chpl_getPrivatizedClass always and loop headers as long as the body of the loop doesn't call functions that modify loop headers.

This got me thinking that'd it be interesting to create an experimental branch that always/forcibly hoists array meta-data and chpl_getPrivatizedClass just so we can do some performance experiments to see potential gains across difference benchmarks. That may also impact how we prioritize this and what direction we end up going for the implementation (i.e. help decide if we're just trying to improve things for prk-stencil, or would a lot of benchmarks benefit from a more rigorous optimization)

**Comment by mppf** [comment]
I think we can hoist chpl_getPrivatizedClass only if the loop body doesn't create a new privatized instance

**Comment by mppf** [comment]
I've created an LLVM bug for the case I was expecting LICM to handle but didn't: https://bugs.llvm.org/show_bug.cgi?id=38933

I'm pretty sure the regression vs. the C backend was due to alias analysis. I've opened issue #11114 to document the problem & discuss what to do about it.

In PR #11113, I adjust chpl_getPrivatizedCopy to generate the load from chpl_privateObjects in Chapel code, so that we get Chapel TBAA metadata. This allows the example at the top of this issue to have competitive performance under --no-local.

**Comment by ronawho** [comment]
> I'm pretty sure the regression vs. the C backend was due to alias analysis.

Cool, that tracks with `The interesting things to note are that clang-included and clang perform as poorly as --llvm and that gcc performs poorly with -fno-strict-aliasing` above

> In PR #11113, I adjust chpl_getPrivatizedCopy to generate the load from chpl_privateObjects in Chapel code, so that we get Chapel TBAA metadata. This allows the example at the top of this issue to have competitive performance under --no-local.

How does performance look for the reproducer with a forall loop and/or the full prk-stencil code?

**Comment by mppf** [comment]
> How does performance look for the reproducer with a forall loop and/or the full prk-stencil code?

I tried with the forall uncommented, GASNet, and 1 locale. --llvm was slightly faster (but probably within noise).

I tried various prk-stencils in the same configuration. "default dist" and "optimized" versions were competitive. The --llvm version for stencil.chpl with StencilDist was 2x slower than C in this configuration (`test/studies/prk/Stencil/stencil.chpl --set iterations=3 --set order=32000 --fast --set useStencilDist=true --llvm`).

We know we can do better at LICM'ing so I still thing it's worth working on that. Wouldn't surprise me if that's the cause of the remaining difference. But, this *immediate issue* is perhaps resolved.
